## Importing required libraries including BeautifulSoup for the webscraping task

### Scraping neighbourhood data from wikipedia and assigning to the variable, _table_

In [1]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

res = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")
soup = BeautifulSoup(res.content,'lxml')
table = soup.find_all('table')[0]

# Saving the table to df variable
df = pd.read_html(str(table))

In [2]:
# Checking the datatype of the variable... It is a list
type(df)

list

In [3]:
# Using a for loop to get a view of the entries in the list

for i in df:
    print (i)

    Postcode           Borough          Neighbourhood
0        M1A      Not assigned           Not assigned
1        M2A      Not assigned           Not assigned
2        M3A        North York              Parkwoods
3        M4A        North York       Victoria Village
4        M5A  Downtown Toronto           Harbourfront
..       ...               ...                    ...
282      M8Z         Etobicoke              Mimico NW
283      M8Z         Etobicoke     The Queensway West
284      M8Z         Etobicoke  Royal York South West
285      M8Z         Etobicoke         South of Bloor
286      M9Z      Not assigned           Not assigned

[287 rows x 3 columns]


In [4]:
# Saving the list data into pandas dataframe

data_df = df[0]
data_df.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

### Checking the number of rows of data

In [5]:
data_df.shape

(287, 3)

### There are currently 287 rows and 3 columns of data. But some data need to be cleaned

Removing the cells with Borough that is "Not assigned", for example, cells M1A and M2A as above

In [6]:
df1 = data_df[data_df['Borough'] != 'Not assigned'].reset_index(drop=True)

In [7]:
df1.head()

Postcode           Borough     Neighbourhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M6A        North York  Lawrence Heights
4      M6A        North York    Lawrence Manor

### Checking the current row count

In [8]:
df1.shape

(210, 3)

### From the current row count, about 77 Boroughs were "Not Assigned" and they were removed

Combining rows with similar Postcode but different Neighbourhood into one row with the neighborhoods separated with a comma

In [9]:
df2 = df1.groupby(['Postcode', 'Borough'])['Neighbourhood'].apply(', '.join).reset_index()

In [10]:
df2.head()

Postcode      Borough                           Neighbourhood
0      M1B  Scarborough                          Rouge, Malvern
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough       Guildwood, Morningside, West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

In [11]:
df2.shape

(103, 3)

There are now 103 rows with unique Postcode and with similar Neighbourhoods separated by comma

## Some cells have Borough but Neighbourhood is Not assigned
Those cells will have their Neighbourbood set to the value of their corresponding Borough


In [12]:
# For instance the Neighbourhood of cell 93 is Not assigned

df2[91:95]

Postcode       Borough                                      Neighbourhood
91      M8Y     Etobicoke  Humber Bay, King's Mill Park, Kingsway Park So...
92      M8Z     Etobicoke  Kingsway Park South West, Mimico NW, The Queen...
93      M9A  Queen's Park                                       Not assigned
94      M9B     Etobicoke  Cloverdale, Islington, Martin Grove, Princess ...

In [13]:
# Importing numpy library

import numpy as np

df2['Neighbourhood'] = np.where(df2['Neighbourhood'] == 'Not assigned', df2['Borough'], df2['Neighbourhood'])
df2[91:95]

Postcode       Borough                                      Neighbourhood
91      M8Y     Etobicoke  Humber Bay, King's Mill Park, Kingsway Park So...
92      M8Z     Etobicoke  Kingsway Park South West, Mimico NW, The Queen...
93      M9A  Queen's Park                                       Queen's Park
94      M9B     Etobicoke  Cloverdale, Islington, Martin Grove, Princess ...

### Neighbourhood of row 93 has been changed to the value of its Borough, that is "Queen's Park"

Finally, the number of rows in our dataset

In [14]:
print(f'The dataframe has {df2.shape[0]} number of rows')

The dataframe has 103 number of rows


## Using coordinates provided in the assignment for the Neighbourhoods of Canada

In [15]:
# Web address of location data csv file
coord_data = 'https://cocl.us/Geospatial_data'

In [16]:
# Reading downloaded csv to a dataframe

df_coord_data = pd.read_csv(coord_data)
df_coord_data.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [17]:
# Joining Latitude and Longitude column in csv file to neighbourhood dataframe

df2['Latitude'] = df_coord_data['Latitude']
df2['Longitude'] = df_coord_data['Longitude']
df2.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

## Boroughs that contains the word Toronto

In [18]:
df3 = df2[df2['Borough'].str.contains('Toronto')].reset_index(drop=True)
df3.head()

Postcode          Borough                   Neighbourhood   Latitude  \
0      M4E     East Toronto                     The Beaches  43.676357   
1      M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2      M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
3      M4M     East Toronto                 Studio District  43.659526   
4      M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [19]:
# Loading geopy library and importing other libraries

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

Getting the coordinates of Toronto 

In [20]:
address = 'Toronto'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


### Creating a map of Toronto

In [21]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood in zip(df3['Latitude'], df3['Longitude'], df3['Borough'], df3['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Using Foursquare API to explore neighbourhoods for segmentation

In [22]:
CLID = '1FNE05U1Y5CS4U4ONNAQ0GGB0JTWKW50WGIN3ZHOVTAJJK3B'
CLSCRT = 'RG51IFNZTERYMK5D2MDEVJKVYCZ11NH4XS1MULKZDUJRIT45'
VERSION = '20200224'

Exploring neighbourhoods in dataframe

In [23]:
df3.loc[0, 'Neighbourhood']

'The Beaches'

Getting the explored neighbourhood's latitude and longitude values

In [24]:
neighbourhood_latitude = df3.loc[0, 'Latitude'] # neighbourhood latitude value
neighbourhood_longitude = df3.loc[0, 'Longitude'] # neighbourhood longitude value

neighbourhood_name = df3.loc[0, 'Neighbourhood'] # neighbourhood name

print(f'Latitude and longitude values of {neighbourhood_name} are {neighbourhood_latitude}, {neighbourhood_longitude}')

Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312


#### Getting the top 100 venues of The Beaches within 500 metres radius

Constructing the url

In [25]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLID, 
    CLSCRT, 
    VERSION, 
    neighbourhood_latitude, 
    neighbourhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=1FNE05U1Y5CS4U4ONNAQ0GGB0JTWKW50WGIN3ZHOVTAJJK3B&client_secret=RG51IFNZTERYMK5D2MDEVJKVYCZ11NH4XS1MULKZDUJRIT45&v=20200224&ll=43.67635739999999,-79.2930312&radius=500&limit=100'

Sending get request and examining results

In [26]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5e548aab9388d7001b456701'},
 'response': {'headerLocation': 'The Beaches',
  'headerFullLocation': 'The Beaches, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 5,
  'suggestedBounds': {'ne': {'lat': 43.680857404499996,
    'lng': -79.28682091449052},
   'sw': {'lat': 43.67185739549999, 'lng': -79.29924148550948}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bd461bc77b29c74a07d9282',
       'name': 'Glen Manor Ravine',
       'location': {'address': 'Glen Manor',
        'crossStreet': 'Queen St.',
        'lat': 43.67682094413784,
        'lng': -79.29394208780985,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.67682094413784,
          'lng': -79.29394208780985}],
        'distanc

From the json data received from Foursquare API, the needed information is in the items key.

Using the _get_category_type_ function from the Foursquare lab:

In [27]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Cleaning the json data and structuring it into a pandas dataframe

In [28]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

print(f'{nearby_venues.shape[0]} venues were returned by Foursquare API.')

5 venues were returned by Foursquare API.


c:\users\femi\appdata\local\programs\python\python37-32\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


5 venues were returned by the Foursquare API

Viewing the venues

In [29]:
nearby_venues

name            categories        lat  \
0                   Glen Manor Ravine                 Trail  43.676821   
1                 Glen Stewart Ravine  Other Great Outdoors  43.676300   
2  The Big Carrot Natural Food Market     Health Food Store  43.678879   
3                 Grover Pub and Grub                   Pub  43.679181   
4                       Upper Beaches          Neighborhood  43.680563   

         lng  
0 -79.293942  
1 -79.294784  
2 -79.297734  
3 -79.297215  
4 -79.292869

## Exploring the neighbourhoods in Toronto

#### A function to repeat the same process for all neighborhoods in Toronto

In [30]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLID, 
            CLSCRT, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Running the created function on each neighbourhood and creating a dataframe called toronto_venues

In [31]:
toronto_venues = getNearbyVenues(names=df3['Neighbourhood'],
                                   latitudes=df3['Latitude'],
                                   longitudes=df3['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North, Forest Hill West
The Annex, North Midtown, Yorkville
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, Exhibition Place, Parkdale Village
High Park, The Junction Sout

#### Checking the size of the dataframe

In [32]:
print(toronto_venues.shape)
toronto_venues.head()

(1718, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0   The Beaches               43.676357               -79.293031   
1   The Beaches               43.676357               -79.293031   
2   The Beaches               43.676357               -79.293031   
3   The Beaches               43.676357               -79.293031   
4   The Beaches               43.676357               -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1                 Glen Stewart Ravine       43.676300       -79.294784   
2  The Big Carrot Natural Food Market       43.678879       -79.297734   
3                 Grover Pub and Grub       43.679181       -79.297215   
4                       Upper Beaches       43.680563       -79.292869   

         Venue Category  
0                 Trail  
1  Other Great Outdoors  
2     Health Food Store  
3                   Pub  
4          Neighborhood

How many venues were returned for each neighbourhood?

In [33]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide, King, Richmond                                               100   
Berczy Park                                                             56   
Brockton, Exhibition Place, Parkdale Village                            22   
Business Reply Mail Processing Centre 969 Eastern                       17   
CN Tower, Bathurst Quay, Island airport, Harbou...                      16   
Cabbagetown, St. James Town                                             44   
Central Bay Street                                                      79   
Chinatown, Grange Park, Kensington Market                               84   
Christie                                                                18   
Church and Wellesley                                                    86   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              38   
Davisville North                                                         8   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                      18   
Design Exchange, Toronto Dominion Centre                               100   
Dovercourt Village, Dufferin                                            18   
First Canadian Place, Underground city                                 100   
Forest Hill North, Forest Hill West                                      4   
Harbord, University of Toronto                                          36   
Harbourfront                                                            48   
Harbourfront East, Toronto Islands, Union Station                      100   
High Park, The Junction South                                           24   
Lawrence Park                                                            4   
Little Portugal, Trinity                                                54   
Moore Park, Summerhill East                                              4   
North Toronto West                                                      21   
Parkdale, Roncesvalles                                                  15   
Queen's Park                                                            40   
Rosedale                                                                 4   
Roselawn                                                                 1   
Runnymede, Swansea                                                      37   
Ryerson, Garden District                                               100   
St. James Town                                                         100   
Stn A PO Boxes 25 The Esplanade                                         93   
Studio District                                                         42   
The Annex, North Midtown, Yorkville                                     21   
The Beaches                                                              5   
The Beaches West, India Bazaar                                          19   
The Danforth West, Riverdale                                            42   

                                                    Neighbourhood Longitude  \
Neighbourhood                                                                 
Adelaide, King, Richmond                                                100   
Berczy Park                                                              56   
Brockton, Exhibition Place, Parkdale Village                             22   
Business Reply Mail Processing Centre 969 Eastern                        17   
CN Tower, Bathurst Quay, Island airport, Harbou...                       16   
Cabbagetown, St. James Town                                              44   
Central Bay Street                                                       79   
Chinatown, Grange Park, Kensington Market                                84   
Christie                                                       

How many unique categories are in the venues returned?

In [34]:
print(f"There are {len(toronto_venues['Venue Category'].unique())} unique categories.")

There are 236 unique categories.


## Analyzing each neighbourhood

In [35]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighbourhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighbourhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0   The Beaches                  0        0                   0             0   
1   The Beaches                  0        0                   0             0   
2   The Beaches                  0        0                   0             0   
3   The Beaches                  0        0                   0             0   
4   The Beaches                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0             0  ...      1              0                              0   
1             0  ...      0              0                              0   
2             0  ...      0              0                              0   
3             0  ...      0              0                              0   
4             0  ...      0              0                              0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Wine Shop  Wings Joint  \
0                 0                      0         0          0            0   
1                 0                      0         0          0            0   
2                 0                      0         0          0            0   
3                 0                      0         0          0            0   
4                 0                      0         0          0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 237 columns]

Size of the new dataframe

In [36]:
toronto_onehot.shape

(1718, 237)

Grouping rows by neighbourhood and by taking the mean of the frequency of occurrence of each category

In [37]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Afghan Restaurant  \
0                            Adelaide, King, Richmond           0.000000   
1                                         Berczy Park           0.000000   
2        Brockton, Exhibition Place, Parkdale Village           0.000000   
3   Business Reply Mail Processing Centre 969 Eastern           0.000000   
4   CN Tower, Bathurst Quay, Island airport, Harbo...           0.000000   
5                         Cabbagetown, St. James Town           0.000000   
6                                  Central Bay Street           0.000000   
7           Chinatown, Grange Park, Kensington Market           0.000000   
8                                            Christie           0.000000   
9                                Church and Wellesley           0.011628   
10                     Commerce Court, Victoria Hotel           0.000000   
11                                         Davisville           0.000000   
12                                   Davisville North           0.000000   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...           0.000000   
14           Design Exchange, Toronto Dominion Centre           0.000000   
15                       Dovercourt Village, Dufferin           0.000000   
16             First Canadian Place, Underground city           0.000000   
17                Forest Hill North, Forest Hill West           0.000000   
18                     Harbord, University of Toronto           0.000000   
19                                       Harbourfront           0.000000   
20  Harbourfront East, Toronto Islands, Union Station           0.000000   
21                      High Park, The Junction South           0.000000   
22                                      Lawrence Park           0.000000   
23                           Little Portugal, Trinity           0.000000   
24                        Moore Park, Summerhill East           0.000000   
25                                 North Toronto West           0.000000   
26                             Parkdale, Roncesvalles           0.000000   
27                                       Queen's Park           0.000000   
28                                           Rosedale           0.000000   
29                                           Roselawn           0.000000   
30                                 Runnymede, Swansea           0.000000   
31                           Ryerson, Garden District           0.000000   
32                                     St. James Town           0.000000   
33                    Stn A PO Boxes 25 The Esplanade           0.000000   
34                                    Studio District           0.000000   
35                The Annex, North Midtown, Yorkville           0.000000   
36                                        The Beaches           0.000000   
37                     The Beaches West, India Bazaar           0.000000   
38                       The Danforth West, Riverdale           0.000000   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0    0.0000              0.0000        0.0000           0.000   
1    0.0000              0.0000        0.0000           0.000   
2    0.0000              0.0000        0.0000           0.000   
3    0.0000              0.0000        0.0000           0.000   
4    0.0625              0.0625        0.0625           0.125   
5    0.0000              0.0000        0.0000           0.000   
6    0.0000              0.0000        0.0000           0.000   
7    0.0000              0.0000        0.0000           0.000   
8    0.0000              0.0000        0.0000           0.000   
9    0.0000              0.0000        0.0000           0.000   
10   0.0000              0.0000        0.0000           0.000   
11   0.0000              0.0000        0.0000           0.000   
12   0.0000              0.0000        0.0000           0.000   
13   0.0000              0.0000        0.0000           0.000   
14   0.0000             

#### Size of the new dataframe

In [38]:
toronto_grouped.shape

(39, 237)

#### Top 5 of the most common venues in each neighbourhood

In [39]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("***** "+hood+" *****")
    tmp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    tmp.columns = ['venue','frequency']
    tmp = tmp.iloc[1:]
    tmp['frequency'] = tmp['frequency'].astype(float)
    tmp = tmp.round({'frequency': 2})
    print(tmp.sort_values('frequency', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

***** Adelaide, King, Richmond *****
             venue  frequency
0      Coffee Shop       0.06
1             Café       0.04
2       Steakhouse       0.04
3              Bar       0.04
4  Thai Restaurant       0.04


***** Berczy Park *****
                venue  frequency
0         Coffee Shop       0.09
1        Cocktail Bar       0.04
2  Seafood Restaurant       0.04
3      Farmers Market       0.04
4         Cheese Shop       0.04


***** Brockton, Exhibition Place, Parkdale Village *****
                venue  frequency
0                Café       0.14
1      Breakfast Spot       0.09
2         Coffee Shop       0.09
3  Italian Restaurant       0.05
4             Stadium       0.05


***** Business Reply Mail Processing Centre 969 Eastern *****
              venue  frequency
0       Yoga Studio       0.06
1     Garden Center       0.06
2              Park       0.06
3        Comic Shop       0.06
4  Recording Studio       0.06


***** CN Tower, Bathurst Quay, Island airport, Har

                venue  frequency
0    Greek Restaurant       0.21
1  Italian Restaurant       0.07
2         Coffee Shop       0.07
3          Restaurant       0.05
4           Bookstore       0.05




#### Putting the data into a pandas dataframe

Function to sort venues in descending order

In [40]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

creating the new dataframe and displaying the top 10 venues for each neighbourhood.

In [41]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village                  Café   
3  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
4  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café            Steakhouse       Thai Restaurant   
1        Farmers Market          Cocktail Bar              Beer Bar   
2           Coffee Shop        Breakfast Spot                   Gym   
3         Auto Workshop            Smoke Shop    Light Rail Station   
4      Airport Terminal   Rental Car Location       Harbor / Marina   

  5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
0                   Bar       Sushi Restaurant            Restaurant   
1            Steakhouse                 Bakery    Seafood Restaurant   
2                Bakery  Performing Arts Venue             Pet Store   
3                   Spa          Burrito Place        Farmers Market   
4         Boat or Ferry                  Plane              Boutique   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0          Concert Hall          Burger Joint         Cosmetics Shop  
1           Cheese Shop                  Café               Pharmacy  
2          Climbing Gym            Restaurant          Burrito Place  
3  Fast Food Restaurant            Restaurant       Recording Studio  
4                   Bar           Coffee Shop       Sculpture Garden

## Neighbourhoods clustering

Run *k*-means to cluster the neighborhood into 5 clusters.

In [42]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[:] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0,
       1, 0, 4, 0, 0, 0, 4, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0])

Creating a new dataframe that includes the cluster as well as the top 10 venues of each neighbourhood

In [43]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df3

# merge toronto_grouped with df3 to add latitude/longitude for each neighbourhood
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Postcode          Borough                   Neighbourhood   Latitude  \
0      M4E     East Toronto                     The Beaches  43.676357   
1      M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2      M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
3      M4M     East Toronto                 Studio District  43.659526   
4      M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue  \
0 -79.293031               0          Neighborhood   
1 -79.352188               0      Greek Restaurant   
2 -79.315572               0        Sandwich Place   
3 -79.340923               0                  Café   
4 -79.388790               1                  Park   

        2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Other Great Outdoors     Health Food Store                   Pub   
1                 Coffee Shop    Italian Restaurant            Restaurant   
2                         Gym         Movie Theater               Brewery   
3                 Coffee Shop             Gastropub               Brewery   
4  Construction & Landscaping           Swim School              Bus Line   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Trail               Dog Run          Dessert Shop   
1  Furniture / Home Store             Bookstore        Ice Cream Shop   
2           Burrito Place                   Pub           Pizza Place   
3                  Bakery    Italian Restaurant   American Restaurant   
4             Yoga Studio    Dim Sum Restaurant           Event Space   

  8th Most Common Venue    9th Most Common Venue       10th Most Common Venue  
0    Dim Sum Restaurant                    Diner               Discount Store  
1               Brewery          Bubble Tea Shop                       Bakery  
2                  Park             Liquor Store             Sushi Restaurant  
3           Yoga Studio  Comfort Food Restaurant           Seafood Restaurant  
4  Ethiopian Restaurant        Electronics Store  Eastern European Restaurant

Visualizing the resulting clusters

In [44]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examining clusters

Examining each cluster and determining the discriminating venue categories that distinguish each cluster. Based on the defining categories.

Cluster 1

In [45]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0       East Toronto               0          Neighborhood   
1       East Toronto               0      Greek Restaurant   
2       East Toronto               0        Sandwich Place   
3       East Toronto               0                  Café   
5    Central Toronto               0                   Gym   
6    Central Toronto               0        Clothing Store   
7    Central Toronto               0           Pizza Place   
9    Central Toronto               0                   Pub   
11  Downtown Toronto               0           Coffee Shop   
12  Downtown Toronto               0           Coffee Shop   
13  Downtown Toronto               0           Coffee Shop   
14  Downtown Toronto               0        Clothing Store   
15  Downtown Toronto               0                  Café   
16  Downtown Toronto               0           Coffee Shop   
17  Downtown Toronto               0           Coffee Shop   
18  Downtown Toronto               0           Coffee Shop   
19  Downtown Toronto               0           Coffee Shop   
20  Downtown Toronto               0           Coffee Shop   
21  Downtown Toronto               0           Coffee Shop   
24   Central Toronto               0        Sandwich Place   
25  Downtown Toronto               0                  Café   
26  Downtown Toronto               0                  Café   
27  Downtown Toronto               0        Airport Lounge   
28  Downtown Toronto               0           Coffee Shop   
29  Downtown Toronto               0           Coffee Shop   
30  Downtown Toronto               0         Grocery Store   
31      West Toronto               0              Pharmacy   
32      West Toronto               0                   Bar   
33      West Toronto               0                  Café   
34      West Toronto               0                   Bar   
35      West Toronto               0        Breakfast Spot   
36      West Toronto               0                  Café   
37  Downtown Toronto               0           Coffee Shop   
38      East Toronto               0           Yoga Studio   

   2nd Most Common Venue   3rd Most Common Venue        4th Most Common Venue  \
0   Other Great Outdoors       Health Food Store                          Pub   
1            Coffee Shop      Italian Restaurant                   Restaurant   
2                    Gym           Movie Theater                      Brewery   
3            Coffee Shop               Gastropub                      Brewery   
5                  Hotel          Breakfast Spot            Food & Drink Shop   
6            Coffee Shop             Yoga Studio         Fast Food Restaurant   
7           Dessert Shop          Sandwich Place                          Gym   
9     Light Rail Station             Coffee Shop                   Bagel Shop   
11                  Park              Restaurant           Italian Restaurant   
12   Japanese Restaurant        Sushi Restaurant                      Gay Bar   
13                Bakery                    Park                          Pub   
14           Coffee Shop                    Café          Japanese Restaurant   
15           Coffee Shop              Restaurant               Clothing Store   
16        Farmers Market            Cocktail Bar                     Beer Bar   
17    Italian Restaurant               Juice Bar               Sandwich Place   
18                  Café              Steakhouse              Thai Restaurant   
19              Aquarium      Italian Restaurant                        Hotel   
20                  Café                   Hotel                   Restaurant   
21                  Café              Restaurant                        Hotel   
24                  Café             Coffee Shop                     Pharmacy   
25            Restaurant                  Bakery                          Bar   
26                   Bar   Vietnamese Restaurant            

In [46]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
4  Central Toronto               1                  Park   

        2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4  Construction & Landscaping           Swim School              Bus Line   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4           Yoga Studio    Dim Sum Restaurant           Event Space   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
4  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant

In [47]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
23  Central Toronto               2         Jewelry Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
23                 Trail    Mexican Restaurant      Sushi Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
23           Yoga Studio    Dim Sum Restaurant           Event Space   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
23  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant

In [48]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22  Central Toronto               3                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22           Yoga Studio          Dessert Shop    Falafel Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22           Event Space  Ethiopian Restaurant     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22  Eastern European Restaurant   Dumpling Restaurant             Donut Shop

In [49]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
8    Central Toronto               4          Tennis Court   
10  Downtown Toronto               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
8                   Park            Playground            Restaurant   
10            Playground                 Trail           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
8                Dog Run          Dessert Shop    Dim Sum Restaurant   
10      Department Store  Ethiopian Restaurant     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
8                         Diner        Discount Store             Donut Shop  
10  Eastern European Restaurant   Dumpling Restaurant             Donut Shop

<hr>

#### End of analysis of Toronto Neighbourhoods

Thanks